### DISCUSSION TITLE
Unit testing generated C code

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#141] [General]
When we build Nelua from source ourselves, we can test it via `make test` which goes through a series of steps to validate its code and libraries.

My question is, how can we unit test the generated C code to verify it produces accurate output?

I'm not talking about sanitizing the code; we already have a flag for that.

**Comment by ghost** [comment]
Generate assembly code of the nelua code
Generate C code of the nelua code

Generate assembly code of the C code generated by nelua
Compare those two assembly code and see if the output is accurate or not

**Reply by stefanos82** [reply]
That's not the unit testing I'm talking about.

See https://github.com/sheredom/utest.h/#design

**Reply by ghost** [reply]
If you are talking about  unit testing the generated C code, you could use a C Unit Testing Library 
See https://github.com/oz123/awesome-c#testing

If you are talking about unit testing the generated C code from nelua code, then never mind this reply

**Comment by edubart** [comment]
> My question is, how can we unit test the generated C code to verify it produces accurate output?

The `make test` has a test to check if the textual output of C generated code is correct, it's the file `spec/cgenerator_spec.nelua`, that file tests most of possible C code generation situations.

And when you are testing the execution of a Nelua program, you are actually testing execution of C code, because Nelua gets translated to C. Thus, if your Nelua code is working correctly, this means that the generated C code is also working correctly.

**Reply by stefanos82** [reply]
Oh wow...I didn't know that, thank you for letting me know @edubart!
